hoot-core/src/main/cpp/hoot/core/algorithms/alpha-shape/AlphaShapeGenerator.cpp

```diff
@@ -106,76 +110,111 @@ std::shared_ptr<Geometry> AlphaShapeGenerator::generateGeometry(OsmMapPtr inputM
+  //  Set working to false after starting all threads so that they end once the queue is empty
+  _working = false;
+  //  Wait on all threads to end
+  for (int i = 0; i < _maxThreads; ++i)
+    threads[i].join();
+
+  if (_stragglers.size() > 0)
   {
-    NodePtr node = it->second;
-    GeometryPtr point(
-      GeometryFactory::getDefaultInstance()->createPoint(
-        geos::geom::Coordinate(node->getX(), node->getY())));
-    if (!geometry->contains(point.get()))
-    {
-      LOG_TRACE(
-        "Point " << point->toString() << " not covered by alpha shape. Buffering and adding it...");
-      point.reset(point->buffer(_buffer));
-      stragglers.push_back(point);
-    }
-    processed++;
+    LOG_DEBUG("Adding " << StringUtils::formatLargeNumber(_stragglers.size())
+              << " point stragglers to alpha shape.");
+    //  Merge the stragglers geometries
+    GeometryPtr merged = GeometryMerger().mergeGeometries(_stragglers, *_geometry->getEnvelopeInternal());
+    //  Join the original geometry with the stragglers geometry
+    _geometry.reset(_geometry->Union(merged.get()));
   }
+  LOG_VART(_geometry->getArea());
+}
 
-  LOG_VARD(processed);
-
-
-  if (stragglers.size() > 0)
+void AlphaShapeGenerator::_coverStragglersWorker()
+{
+  //  Make a thread local copy of the geometry because Geometry::contains() is not thread-safe
+  GeometryPtr copy(_geometry->clone());
+  //  Set the initial stack size to int max to enter the loop once and check actual stack size
+  int stack_size = std::numeric_limits<int>::max();
+  //  Keep looping while there is work on the stack
+  while (_working || stack_size > 0)
   {
-    LOG_DEBUG("Adding " << stragglers.size() << " point stragglers to alpha shape.");
-    //  Merge the stragglers geometries
-    GeometryPtr merged = GeometryUtils::mergeGeometries(stragglers, *geometry->getEnvelopeInternal());
-    //  Join the original geometry with the stragglers geometry
-    geometry.reset(geometry->Union(merged.get()));
+    //  Get the top node off of the stack
+    _nodesMutex.lock();
+    NodePtr node;
+    stack_size = _nodes.size();
+    if (stack_size > 0)
+    {
+      node = _nodes.top();
+      _nodes.pop();
+      --stack_size;
+    }
+    _nodesMutex.unlock();
+    //  Where the node is valid check ::contains()
+    if (node)
+    {
+      GeometryPtr point(
+        GeometryFactory::getDefaultInstance()->createPoint(
+          geos::geom::Coordinate(node->getX(), node->getY())));
+      //  Geometry::contains() is a time consuming function call (and not thread safe
+      //  so all threads need their own `copy`)
+      if (!copy->contains(point.get()))
+      {
+        LOG_TRACE(
+          "Point " << point->toString() << " not covered by alpha shape. Buffering and adding it...");
+        point.reset(point->buffer(_buffer));
+        _stragglersMutex.lock();
+        _stragglers.push_back(point);
+        _stragglersMutex.unlock();
+      }
```

Comment:
This is the whole point to the thread.  Calling `Geometry::contains` on every single node is quite a long process but it is independent of any other node so they can all be done in parallel.